sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubqueryBroadcastExec.scala

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.execution
+
+import scala.collection.JavaConverters.asScalaIteratorConverter
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration.Duration
+
+import ai.rapids.cudf.NvtxColor
+import com.nvidia.spark.rapids.{GpuColumnVector, GpuExec, GpuMetric, NvtxWithMetrics}
+import com.nvidia.spark.rapids.GpuMetric.{COLLECT_TIME, DESCRIPTION_COLLECT_TIME, ESSENTIAL_LEVEL}
+import com.nvidia.spark.rapids.shims.v2.ShimUnaryExecNode
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Expression, UnsafeProjection}
+import org.apache.spark.sql.execution.{BaseSubqueryExec, SparkPlan, SQLExecution}
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.util.ThreadUtils
+
+
+case class GpuSubqueryBroadcastExec(
+    name: String,
+    index: Int,
+    buildKeys: Seq[Expression],
+    child: SparkPlan) extends BaseSubqueryExec with GpuExec with ShimUnaryExecNode {
+
+  override lazy val additionalMetrics: Map[String, GpuMetric] = Map(
+    "dataSize" -> createSizeMetric(ESSENTIAL_LEVEL, "data size"),
+    COLLECT_TIME -> createNanoTimingMetric(ESSENTIAL_LEVEL, DESCRIPTION_COLLECT_TIME))
+
+  @transient
+  private lazy val relationFuture: Future[Array[InternalRow]] = {
+    // relationFuture is used in "doExecute". Therefore we can get the execution id correctly here.
+    val executionId = sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
+    val collectTime = gpuLongMetric(COLLECT_TIME)
+
+    Future {
+      // This will run in another thread. Set the execution id so that we can connect these jobs
+      // with the correct execution.
+      SQLExecution.withExecutionId(sparkSession, executionId) {
+        withResource(new NvtxWithMetrics("broadcast collect", NvtxColor.GREEN,
+          collectTime)) { _ =>
+          val serBatch = child.executeBroadcast[SerializeConcatHostBuffersDeserializeBatch]().value
```

Comment:
This is running on the driver, but assumes that it has access to a GPU. It does not. We have to do any/all transformation on the CPU.

Reply:
Hi @revans2, I overlooked the access of GPU at the first time. I refactored the implementation. For now, the GpuSubqueryBroadcast is entirely on host.

Reply:
Could we please try to test this on the YARN cluster or some place where there is no GPU and ideally no CUDA on the nodes that are running the driver? I just want to be sure that we don't accidentally initialize the CUDA context when we try to touch the HostColumnVectors.  I think in the other places we only touched buffers.

Reply:
Hi @revans2, I ran a spark yarn test with a pure CPU driver container. The `GpuSubqueryBroadcast` didn't throw any exception.